### PR TITLE
[BitMex] Authenticated Channels

### PR DESCRIPTION
**Pull request by mdvx**
Bitmex use ExchangeSpecification to select Sandbox, ApiKey and ApiSecret

expose getStreamingService() for Bitmex and CoinbasePro

Create to resubscribe channels - this is also called manually after the socket connects  (would be nice to do that automatically)
    @Override  // called by NettyStreamingService.resubscribeChannels
    public String BitmexStreamingService.getAuthenticateMessage() throws IOException;

this gives access to the authenticated channels

### CODE REVIEW

**Comment by Flemingjp** [comment]
#268 merged. @mdvx any comments on @bagerwithagun comments?

File: pom.xml [hunk 1]
```diff
@@ -80,7 +80,7 @@
     </repositories>
 
     <properties>
-        <xchange.version>4.3.13</xchange.version>
+        <xchange.version>4.3.14-SNAPSHOT</xchange.version>
```
[the remark this answers is not in the document]
Comment (mdvx) [reply pom.xml]: will do

File: xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingExchange.java [hunk 2]
```diff
@@ -65,4 +70,8 @@ public boolean isAlive() {
 
     @Override
     public void useCompressedMessages(boolean compressedMessages) { streamingService.useCompressedMessages(compressedMessages); }
+
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingExchange.java]: Yes, I agree, I wanted to open this up for discussuion

File: xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingService.java [hunk 3]
```diff
@@ -1,41 +1,56 @@
...
+    }
+    public BitmexStreamingService(String apiUrl, ExchangeSpecification exchangeSpecification) {
+        super(apiUrl, Integer.MAX_VALUE);
+        this.exchangeSpecification = exchangeSpecification;
     }
 
-	 @Override
+    @Override
     protected void handleMessage(JsonNode message) {
+//        if (message.has("info") && message.get("info").asText().startsWith("Welcome ")) {
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingService.java]: yes

File: xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingService.java [hunk 4]
```diff
@@ -73,4 +93,30 @@ public String getUnsubscribeMessage(String channelName) throws IOException {
...
+
+        SynchronizedValueFactory<Long> nonceFactory = new ExpirationTimeFactory(30);
+
+        long nonce = nonceFactory.createValue();
+        String payload = "GET/realtime" + nonce;
+        String digestString = bitmexDigest.digestString(payload);
+
+        BitmexWebSocketSubscriptionMessage subscribeMessage =
+                new BitmexWebSocketSubscriptionMessage("authKeyExpires",
+                new Object[]{exchangeSpecification.getApiKey(), nonce, digestString});
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-bitmex/src/main/java/info/bitrich/xchangestream/bitmex/BitmexStreamingService.java]: yes, i first iteration passed a digest object, but a string is suffecient now

**Comment by mdvx** [comment]
I am a little stuck here, I am not sure how to pull the PR down onto a clean checkout of develop branch, so I can make the requested changes.   

Are there some general instructions somewhere.

I am using git (normal, tortoise, InteliJ) on windows.

**Comment by badgerwithagun** [comment]
@mdvx : I'm not an IntelliJ user myself, but the sequence of actions you need is:

1. `git fetch upstream` - this updates your local version of the bitrich-info github change history, assuming your remote is called `upstream`.  This is the default, but it might not be called that on your machine.
1. `git merge upstream/develop` - this merges the changes in the bitrich-info github repo into your local repo

You can then `push` your changes up to your github repo.

More here: https://help.github.com/articles/syncing-a-fork/

**Comment by badgerwithagun** [comment]
See #274 - if this PR can be brought in line with the approach detailed there (similar to #246 and #273) that would be great.  If not, I'll pick this up and unify it with #244 when I've finished work on Binance, Coinbase Pro and Bitfinex.

**Comment by mdvx** [comment]
will return to BitMEX, using trunc version

**Pull request closed without merging**